velox/expression/SignatureBinder.cpp

```diff
@@ -122,90 +118,120 @@ bool SignatureBinder::tryBind() {
   return true;
 }
 
+bool SignatureBinder::tryAndSetIntegerParameter(
+    const std::string& parameterName,
+    int value) {
+  auto it = integerParameters_.find(parameterName);
+  // Return false if the parameter is not found.
+  if (it == integerParameters_.end()) {
+    return false;
+  }
+  // Return false if the parameter is found with a different value.
+  if (it->second.has_value() && it->second.value() != value) {
+    return false;
+  }
+  it->second = value;
+  return true;
+}
+
+bool SignatureBinder::tryBindIntegerParameters(
+    const std::vector<exec::TypeSignature>& parameters,
+    const TypePtr& actualType) {
+  // Only decimal types are supported.
+  if (!(actualType->isShortDecimal() || actualType->isLongDecimal())) {
+    return false;
+  }
+
+  VELOX_CHECK_EQ(parameters.size(), 2);
+  const auto& [precision, scale] = getDecimalPrecisionScale(*actualType);
+  return tryAndSetIntegerParameter(parameters[0].baseName(), precision) &&
+      tryAndSetIntegerParameter(parameters[1].baseName(), scale);
+}
+
 bool SignatureBinder::tryBind(
     const exec::TypeSignature& typeSignature,
     const TypePtr& actualType) {
   if (isAny(typeSignature)) {
     return true;
   }
 
-  auto it = bindings_.find(typeSignature.baseType());
-  if (it == bindings_.end()) {
-    // concrete type
-    auto typeName = boost::algorithm::to_upper_copy(typeSignature.baseType());
-    if (isDecimalName(typeName)) {
-      VELOX_USER_FAIL("Use 'DECIMAL' in the signature.");
-    }
-    if (isDecimalKind(actualType->kind()) && isCommonDecimalName(typeName)) {
-      const auto& variables = typeSignature.variables();
-      VELOX_CHECK_EQ(variables.size(), 2);
-      const auto& [precision, scale] = getDecimalPrecisionScale(*actualType);
-      variables_.emplace(variables[0], precision);
-      variables_.emplace(variables[1], scale);
-      return true;
-    }
+  auto isConcreteType =
+      (typeParameters_.count(typeSignature.baseName()) == 0) &&
+      (integerParameters_.count(typeSignature.baseName()) == 0);
+  if (isConcreteType) {
```

Comment:
This code pattern is also used in tryResolveType(). Maybe extract it into a helper function?